In [1]:
# Class: COEN 140 Lab Machine Learning/Data Mining
# Name: Matthew Davenport
# Date: 9/25/2022
# Title: Lab 5 – Crime predictions using linear and ridge regression without python libraries
# Description: This program will perform clssifications
#           on crime datasets
#           without using python ML libraries
#           implementing linear and ridge regression

import numpy as np
import math
import time

In [44]:
# import data from txt file for test data and training data
import urllib.request
import pandas as pd

test_data = urllib.request.urlopen('https://www.cse.scu.edu/~yfang/coen140/crime-test.txt')
train_data = urllib.request.urlopen('https://www.cse.scu.edu/~yfang/coen140/crime-train.txt')

train_data = pd.read_csv(train_data, delimiter='\t')
test_data = pd.read_csv(test_data, delimiter='\t')
print(train_data)
print(test_data)


      ViolentCrimesPerPop  population  householdsize  agePct12t21  \
0                    0.67       -0.45          -1.85        -1.06   
1                    0.43       -0.45          -0.27        -0.22   
2                    0.12       -0.14           1.87         0.55   
3                    0.03       -0.38           0.53        -0.28   
4                    0.14       -0.30          -1.12        -0.74   
...                   ...         ...            ...          ...   
1590                 0.09       -0.38          -0.39         0.04   
1591                 0.45       -0.06           3.03         1.71   
1592                 0.23        0.81          -0.57        -0.48   
1593                 0.19        0.18           0.28         1.00   
1594                 0.48        1.12           1.93         0.49   

      agePct12t29  agePct16t24  agePct65up  numbUrban  pctUrban  medIncome  \
0            0.67         0.08       -0.85      -0.34      0.68      -0.24   
1           -0.

In [45]:
# format data (separations for X and y, and adding dummy feature)
y_test = test_data.iloc[:, 0]
y_train = train_data.iloc[:, 0]
X_test = test_data.iloc[: , 1:]
X_train = train_data.iloc[: , 1:]

# correct the type to float64
X_train = np.float64(X_train)
X_test = np.float64(X_test)
y_train = np.float64(y_train)
y_test = np.float64(y_test)

# add dummy feature to X training and testing matrix
X_train = np.column_stack((X_train, np.ones(len(X_train))))
X_test = np.column_stack((X_test, np.ones(len(X_test))))


# df = df.iloc[: , 1:] dataframe keeping everything except first column 
# dataframe.iloc[:, 0] dataframe extracting only first column

print(X_test)
print(y_test)

[[-0.14  0.35 -0.41 ...  0.65 -0.39  1.  ]
 [ 0.02 -0.45 -0.22 ... -0.66 -0.39  1.  ]
 [-0.45  0.28 -0.16 ... -0.66 -0.39  1.  ]
 ...
 [-0.38  1.99  1.07 ... -0.57 -0.39  1.  ]
 [-0.38  0.04 -0.22 ... -0.27 -0.39  1.  ]
 [ 0.02 -0.57 -0.48 ... -0.14 -0.39  1.  ]]
[0.08 0.22 0.06 0.16 0.15 0.28 0.63 0.16 0.25 1.   0.04 0.05 0.86 0.08
 0.09 0.17 0.12 0.12 1.   0.16 0.4  0.12 0.07 0.4  0.06 0.21 0.04 0.45
 0.02 0.28 0.38 0.22 0.07 0.06 0.16 0.21 1.   0.19 0.26 0.1  0.23 0.09
 0.55 1.   0.09 0.41 1.   0.05 0.02 0.09 0.11 0.36 0.03 0.63 0.04 0.02
 0.08 0.27 0.52 0.27 0.31 0.06 0.1  0.22 0.6  0.49 0.4  0.02 0.21 0.2
 1.   0.5  0.15 0.31 0.22 0.16 0.16 0.32 0.21 0.09 0.09 0.02 0.05 0.15
 0.11 0.1  0.06 0.11 0.01 0.1  0.18 0.85 0.05 0.12 0.54 0.06 0.04 0.18
 0.44 0.03 0.14 0.03 0.29 0.14 0.71 0.06 0.35 0.07 0.2  0.3  0.05 0.05
 0.26 0.62 0.11 0.03 0.17 0.25 0.34 0.39 0.13 0.19 0.11 0.06 0.08 0.01
 0.04 0.25 0.08 0.31 1.   0.73 0.18 0.37 1.   0.37 0.04 0.05 0.04 0.2
 0.33 0.06 0.03 0.29 1.   0.

Part 1: Implement linear regression model
Compute the RMSE value on the training data 
and test data, respectively. Report both RSME values. 
Provide a function that, given an NxP numpy array of data points, returns an Nx1 
array of predicted outputs. 

Problem1: Given a sample, using linear regression, return a vector of predicted outputs


In [46]:
def problem2(samples):
    w = np.dot((np.linalg.inv(np.dot(X_train.T, X_train))), np.dot(X_train.T, y_train))
    predicted_outputs = []
    for X in samples: 
        predicted_outputs.append(np.dot(X.T, w))
    return predicted_outputs 

RMSE: Given a list of predicted outputs and correct outputs, return the root mean square error 

In [47]:
def RMSE(predicted_outputs, correct_outputs):
    total = 0
    differences = []
    m = len(predicted_outputs)
    differences = (predicted_outputs - correct_outputs) ** 2
    for difference in differences: 
        total += difference
    error = math.sqrt(total/m)
    return error

Now apply problem 1 linear regression to the dataset

In [48]:
# compute predictions for training and testing data
linear_training = problem1(X_train)
linear_testing = problem1(X_test)

# compute the root mean square error for both testing and training predictions
linear_training_rmse = RMSE(linear_training, y_train)
linear_testing_rmse = RMSE(linear_testing, y_test)

# display results
print("RMSE for training: ", linear_training_rmse)
print("RMSE for testing: ", linear_testing_rmse)


RMSE for training:  0.127689674217622
RMSE for testing:  0.14583464490949688


Part 2: Implement ridge regression model with lambda = 100. Compute the RMSE value on the training data and test data, respectively. Report both RSME values. Provide a function that, given an NxP numpy array of data points, returns an Nx1 array of predicted outputs.

In [49]:
def problem2(samples):
    reg_lambda = 100
    predicted_outputs = []
    w = np.dot((np.linalg.inv((np.dot(X_train.T, X_train)) + reg_lambda * np.identity(len(X_train.T)))), \
               np.dot(X_train.T, y_train))
    for X in samples:
        predicted_outputs.append(np.dot(X.T, w))
        
    return predicted_outputs

Now apply ridge regression to the test and train data sets and display the RMSE for both

In [50]:
# compute predictions for training and testing data
ridge_training = problem2(X_train)
ridge_testing = problem2(X_test)

# compute the root mean square error for both testing and training predictions
ridge_training_rmse = RMSE(ridge_training, y_train)
ridge_testing_rmse = RMSE(ridge_testing, y_test)

# display results
print("RMSE for training: ", ridge_training_rmse)
print("RMSE for testing: ", ridge_testing_rmse)

RMSE for training:  0.13134320424615795
RMSE for testing:  0.1476569846852611


Results: 

Using linear regression:
RMSE for training predictions: 0.127689674217622 = 12.77 %
RMSE for testing predictions: 0.14583464490949688 = 14.58 %

Using ridge regression: 
RMSE for training predictions: 0.13134320424615795 = 13.13 %
RMSE for testing predictions: 0.1476569846852611 = 14.77 %

Using a lambda value of 100 with ridge regression produces a slightly higher root mean square error than just linear regression on the dataset. 